xarray/core/variable.py

```diff
@@ -1212,35 +1212,94 @@ def __setitem__(self, key, value):
         raise TypeError('%s values cannot be modified' % type(self).__name__)
 
     @classmethod
-    def concat(cls, variables, dim='concat_dim', positions=None,
-               shortcut=False):
-        """Specialized version of Variable.concat for IndexVariable objects.
-
-        This exists because we want to avoid converting Index objects to NumPy
-        arrays, if possible.
+    def concat_numpy(cls, variables, positions=None):
         """
-        if not isinstance(dim, basestring):
-            dim, = dim.dims
+        Concatenates variables. Works for variables whose dtype is
+        different from numpy.object. If variables' dtype is numpy.object
+        it throws TypeError and "concat" function will use
+        concat_pandas function
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+        variable_type_set = set(map(lambda v: type(v.data), variables))
 
-        variables = list(variables)
-        first_var = variables[0]
+        if len(variable_type_set) > 1:
+            raise TypeError('Trying to concatenate variables of '
+                            'different types')
 
-        if any(not isinstance(v, cls) for v in variables):
-            raise TypeError('IndexVariable.concat requires that all input '
-                            'variables be IndexVariable objects')
+        variable_type = list(variable_type_set)[0]
+        if not variable_type == np.ndarray:
+            raise TypeError('Can only concatenate variables whose '
+                            '_data member is ndarray')
```

Comment:
This is already guaranteed by xarray's data model, so this should either be an `assert` or you should skip it entirely.